File: 2.2Museo_Java/Clases_Museo/Catalogo.java
package Clases_Museo;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.stream.Collectors;

/**
 * Gestiona el conjunto de obras de arte registradas en el museo.
 *
 * Permite agregar, eliminar y consultar obras por distintos
 * criterios: autor, periodo, estado, sala y tipo.
 */
public class Catalogo {

    // Atributo privado
    private List<Obra> obras;

    // --- Constructor ---

    /** Inicializa el catalogo con una lista de obras vacia. */
    public Catalogo() {
        this.obras = new ArrayList<>();
    }

    // --- Getter ---

    /** @return Una copia de la lista de obras registradas. */
    public List<Obra> getObras() { return new ArrayList<>(obras); }

    // --- Metodos de registro ---

    /**
     * Agrega una obra al catalogo si no estaba registrada.
     *
     * @param obra Obra a agregar.
     * @throws IllegalArgumentException si la obra ya esta en el catalogo.
     */
    public void agregarObra(Obra obra) {
        if (obras.contains(obra)) {
            throw new IllegalArgumentException(
                "La obra '" + obra.getTitulo() + "' ya esta registrada en el catalogo.");
        }
        obras.add(obra);
    }

    /**
     * Elimina una obra del catalogo.
     *
     * @param obra Obra a eliminar.
     * @throws IllegalArgumentException si la obra no esta en el catalogo.
     */
    public void eliminarObra(Obra obra) {
        if (!obras.remove(obra)) {
            throw new IllegalArgumentException(
                "La obra '" + obra.getTitulo() + "' no esta en el catalogo.");
        }
    }

    // --- Consultas ---

    /**
     * Retorna las obras de un autor dado (busqueda insensible a mayusculas).
     *
     * @param autor Nombre del autor a buscar.
     * @return Lista de obras del autor.
     */
    public List<Obra> buscarPorAutor(String autor) {
        String autorLower = autor.toLowerCase();
        return obras.stream()
            .filter(o -> o.getAutor().toLowerCase().equals(autorLower))
            .collect(Collectors.toList());
    }

    /**
     * Retorna las obras correspondientes a un periodo artistico.
     *
     * @param periodo Periodo historico a filtrar.
     * @return Lista de obras del periodo.
     */
    public List<Obra> buscarPorPeriodo(Obra.Periodo periodo) {
        return obras.stream()
            .filter(o -> o.getPeriodo() == periodo)
            .collect(Collectors.toList());
    }

    /**
     * Retorna las obras que se encuentran en un estado determinado.
     *
     * @param estado Estado de la obra a filtrar.
     * @return Lista de obras en ese estado.
     */
    public List<Obra> buscarPorEstado(Obra.EstadoObra estado) {
        return obras.stream()
            .filter(o -> o.getEstado() == estado)
            .collect(Collectors.toList());
    }

    /**
     * Retorna las obras asignadas a una sala, ordenadas por titulo.
     *
     * @param sala Sala por la que filtrar.
     * @return Lista de obras en la sala, ordenadas alfabeticamente.
     */
    public List<Obra> buscarPorSala(Sala sala) {
        return obras.stream()
            .filter(o -> sala.equals(o.getSala()))
            .sorted(Comparator.comparing(Obra::getTitulo))
            .collect(Collectors.toList());
    }

    /**
     * Retorna las obras de un tipo concreto: Cuadro, Escultura u Otra.
     *
     * @param tipo Tipo de obra a filtrar.
     * @return Lista de obras del tipo indicado.
     */
    public List<Obra> buscarPorTipo(String tipo) {
        return obras.stream()
            .filter(o -> o.obtenerTipo().equals(tipo))
            .collect(Collectors.toList());
    }

    @Override
    public String toString() {
        if (obras.isEmpty()) {
            return "El catalogo esta vacio.";
        }
        StringBuilder sb = new StringBuilder("Catalogo (" + obras.size() + " obras):");
        for (Obra obra : obras) {
            sb.append("\n  - ").append(obra);
        }
        return sb.toString();
    }
}

File: 1.2Museo_Python/Clases/Catalogo.py
from __future__ import annotations

import os
import sys

sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), "..")))

from Clases.Obra import EstadoObra, Periodo


class Catalogo:
    """Gestiona el conjunto de obras de arte registradas en el museo."""

    def __init__(self) -> None:
        """Inicializa el catálogo con una lista de obras vacía."""
        self._obras: list = []

    # --- getters ---

    @property
    def obras(self) -> list:
        """Retorna una copia de la lista de obras registradas."""
        return list(self._obras)

# --------------------Metodos--------------------------------------------------

    def agregar_obra(self, obra) -> None:
        """Añade una obra al catálogo si no estaba registrada."""
        if obra in self._obras:
            raise ValueError(f"La obra '{obra.titulo}' ya está registrada en el catálogo.")
        self._obras.append(obra)

    def eliminar_obra(self, obra) -> None:
        """Elimina una obra del catálogo."""
        if obra not in self._obras:
            raise ValueError(f"La obra '{obra.titulo}' no está en el catálogo.")
        self._obras.remove(obra)

# --- Consultas -------------------------------------------------------------

    def buscar_por_autor(self, autor: str) -> list:
        """Retorna las obras de un autor dado (búsqueda insensible a mayúsculas)."""
        autor_lower = autor.lower()
        return [o for o in self._obras if o.autor.lower() == autor_lower]

    def buscar_por_periodo(self, periodo: Periodo) -> list:
        """Retorna las obras correspondientes a un periodo artístico."""
        return [o for o in self._obras if o.periodo == periodo]

    def buscar_por_estado(self, estado: EstadoObra) -> list:
        """Retorna las obras que se encuentran en un estado determinado."""
        return [o for o in self._obras if o.estado == estado]

    def buscar_por_sala(self, sala) -> list:
        """Retorna las obras asignadas a una sala, ordenadas por título."""
        return sorted(
            [o for o in self._obras if o.sala == sala],
            key=lambda o: o.titulo,
        )

    def buscar_por_tipo(self, tipo: str) -> list:
        """Retorna las obras de un tipo concreto: 'Cuadro', 'Escultura' u 'Otra'."""
        return [o for o in self._obras if o.obtener_tipo() == tipo]



    def __str__(self) -> str:
        if not self._obras:
            return "El catálogo está vacío."
        lineas = [f"Catálogo ({len(self._obras)} obras):"]
        for obra in self._obras:
            lineas.append(f"  - {obra}")
        return "\n".join(lineas)


# ---------------------------------------------------------------------------
# Prueba
# ---------------------------------------------------------------------------

if __name__ == "__main__":
    import io
    from datetime import date

    from Clases.Cuadro import Cuadro, EstiloCuadro, TecnicaCuadro
    from Clases.Escultura import Escultura, MaterialEscultura
    from Clases.Obra import Periodo

    # Silenciar creación de objetos
    _stdout = sys.stdout
    sys.stdout = io.StringIO()

    catalogo = Catalogo()

    cuadro1 = Cuadro(
        "La noche estrellada", "Van Gogh", Periodo.IMPRESIONISMO,
        date(1889, 6, 1), date(1941, 1, 1),
        EstiloCuadro.POSTIMPRESIONISMO, TecnicaCuadro.OLEO,
    )
    cuadro2 = Cuadro(
        "Las Meninas", "Diego Velázquez", Periodo.BARROCO,
        date(1656, 1, 1), date(1819, 11, 19),
        EstiloCuadro.BARROCO, TecnicaCuadro.OLEO,
    )
    escultura = Escultura(
        "El pensador", "Auguste Rodin", Periodo.IMPRESIONISMO,
        date(1904, 1, 1), date(1950, 3, 15),
        MaterialEscultura.BRONCE,
    )

    catalogo.agregar_obra(cuadro1)
    catalogo.agregar_obra(cuadro2)
    catalogo.agregar_obra(escultura)

    sys.stdout = _stdout

    print("=== Prueba de Catálogo ===")
    print(catalogo)

    print("\n--- Menú de búsqueda ---")
    print("1. Buscar por autor")
    print("2. Buscar por periodo")
    print("3. Buscar por tipo")
    opcion = input("Elige una opción: ")

    if opcion == "1":
        autor = input("Autor: ")
        resultados = catalogo.buscar_por_autor(autor)
    elif opcion == "2":
        print("Periodos disponibles:")
        for p in Periodo:
            print(f"  {p.name}: {p.value}")
        nombre_periodo = input("Nombre del periodo (ej. BARROCO): ").upper()
        resultados = catalogo.buscar_por_periodo(Periodo[nombre_periodo])
    elif opcion == "3":
        tipo = input("Tipo (Cuadro / Escultura / Otra): ")
        resultados = catalogo.buscar_por_tipo(tipo)
    else:
        resultados = []
        print("Opción no válida.")

    if resultados:
        print(f"\nResultados ({len(resultados)}):")
        for r in resultados:
            print(f"  - {r}")
    elif opcion in ("1", "2", "3"):
        print("No se encontraron obras.")
